### PR TITLE
Gr740

### PR DESCRIPTION
**Pull request by timcanham**
Updates to support VxWorks 7 and  fix to Os::Log.

### CODE REVIEW

File: cmake/toolchain/gr740-vxworks7.cmake [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# Set system name
+set(CMAKE_SYSTEM_NAME VxWorks)
+set(CMAKE_SYSTEM_VERSION 7.0)
+set(CMAKE_SYSTEM_PROCESSOR gr740)
+
+include_directories(SYSTEM "/opt/tools/sabertooth/sabertooth-bsp/LeonVSB/krnl/h/public")
+include_directories(SYSTEM "/opt/tools/sabertooth/sabertooth-bsp/LeonVSB/krnl/h/system")
+include_directories("/opt/tools/sabertooth/sabertooth-bsp/LeonVSB/share/h")
```
Comment (kevin-f-ortega) [inline cmake/toolchain/gr740-vxworks7.cmake:8]: These lines should use `%(VSB_HOME)s`

File: cmake/toolchain/gr740-vxworks7.cmake [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+add_definitions(-DBUILD_SABERTOOTH)
+
+set(WINDRIVER_COMPILER_ROOT "/opt/tools/sabertooth/Windriver-7-SR0620/compilers/gaisler-gnu-7.2.0.1/x86_64-linux2")
```
Comment (kevin-f-ortega) [inline cmake/toolchain/gr740-vxworks7.cmake:12]: This line should use `%(WIND_HOME)s`

File: cmake/env/VxWorks-GR740.ini [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+MUNCH=tclsh %(WIND_BASE)s/build/mk/tcl/munch.tcl -c sparc -tags %(VSB_HOME)s/krnl/tags/dkm.tags
+CC_MUNCH=%(WIND_COMP)s/ccsparc -fdollars-in-identifiers -w
+CFLAGS=-O3 -DTOOL_FAMILY=gnu -DTOOL=gnu -Wall -Wextra -fno-builtin -Wno-unused-parameter -Wno-long-long -Wconversion -Wno-sign-conversion -ansi -std=c99 -fno-cond-mismatch -pedantic -Werror-implicit-function-declaration -Wstrict-prototypes -DFPRIME -DBUILD_SABERTOOTH -DTGT_OS_TYPE_VXWORKS -DTGT_OS_TYPE_VXWORKS7 -D__vxworks -D__VXWORKS__ -D__ELF__ -D_HAVE_TOOL_XTORS -D_VSB_CONFIG_FILE=\"/opt/tools/sabertooth/sabertooth-bsp/LeonVSB/h/config/vsbConfig.h\" -D_WRS_KERNEL -ftls-model=local-exec -mcpu=leon3 -DCPU=_VX_SPARC -DCPU_VARIANT=_sparcleon -fno-zero-initialized-in-bss -D_WRS_VX_SMP -D_WRS_CONFIG_SMP -fno-builtin -fno-strict-aliasing -D_USE_INIT_ARRAY -isystem%(VSB_HOME)s/krnl/h/public -isystem%(VSB_HOME)s/krnl/h/system -I%(VSB_HOME)s/share/h
+COMPILE_ONLY=-c
+COMPILE_TO=-o
+LINK_BIN_POST=%(WIND_COMP)s/ldsparc
+LINK_BIN_POST_FLAGS=-X -r --eh-frame-hdr --force-group-allocation -T %(VSB_HOME)s/krnl/ldscripts/dkm.ld
+LINK_BIN_POST_TO=-o
+SYMBOL_CHECK=echo
+SYMBOL_CHECK_FILE=/opt/tools/sabertooth/kernels/vxWorks7-09-30-2020
```
Comment (kevin-f-ortega) [inline cmake/env/VxWorks-GR740.ini:21]: Should use `%(WIND_TOOLS_HOME)s`

File: cmake/toolchain/gr740-vxworks7.cmake [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+
+set(COMPILER_COMMON_FLAGS
+    "-DTGT_OS_TYPE_VXWORKS \
+    -DTGT_OS_TYPE_VXWORKS7 \
+    -D__vxworks \
+    -D__VXWORKS__ \
+    -D__ELF__  \
+    -D_HAVE_TOOL_XTORS \
+    -D_VSB_CONFIG_FILE=\\\"/opt/tools/sabertooth/sabertooth-bsp/LeonVSB/h/config/vsbConfig.h\\\" \
```
Comment (kevin-f-ortega) [inline cmake/toolchain/gr740-vxworks7.cmake:43]: Can this line use `%(VSB_HOME)s` ? If so, then this line should use that.

File: fprime-vxworks/VxWorks/Log.cpp [hunk 5]
```diff
@@ -25,6 +30,15 @@ namespace Os {
         POINTER_CAST a8,
         POINTER_CAST a9
     ) {
-        ::logMsg(const_cast<char*>(fmt), a1, a2, a3, a4, a5, a6);
+
+        // copy the possibly transient string to a circular buffer
+        // this relies on the VxWorks log task to keep up with
+        // the circular buffer
+
+        strncpy(logBuffers[logEntry],fmt,FW_LOG_TEXT_BUFFER_SIZE);
```
Comment (kevin-f-ortega) [inline fprime-vxworks/VxWorks/Log.cpp:38]: Can you please void out the return by changing `strncpy` to `(void)strncpy` ?

**Review by kevin-f-ortega — COMMENTED** [review]
Looks good. Just a few minor comments.

File: fprime-vxworks/VxWorks/Log.cpp (hunk 5 above)
Comment (timcanham) [reply fprime-vxworks/VxWorks/Log.cpp]: Done!

File: cmake/toolchain/gr740-vxworks7.cmake (hunk 1 above)
Comment (timcanham) [reply cmake/toolchain/gr740-vxworks7.cmake:8]: The `%X` variables are only available when the make is run (`fprime-util build`), not during the cmake generate stage (`fprime-util generate`), so I can't use them.

File: cmake/toolchain/gr740-vxworks7.cmake (hunk 2 above)
Comment (timcanham) [reply cmake/toolchain/gr740-vxworks7.cmake:12]: See above...

File: cmake/env/VxWorks-GR740.ini (hunk 3 above)
Comment (timcanham) [reply cmake/env/VxWorks-GR740.ini:21]: See above...

File: cmake/toolchain/gr740-vxworks7.cmake (hunk 4 above)
Comment (timcanham) [reply cmake/toolchain/gr740-vxworks7.cmake:43]: See above...

File: cmake/toolchain/gr740-vxworks7.cmake (hunk 1 above)
Comment (kevin-f-ortega) [reply cmake/toolchain/gr740-vxworks7.cmake:8]: Ah, okay.

File: fprime-vxworks/VxWorks/Log.cpp [hunk 6]
```diff
@@ -11,6 +12,10 @@ namespace Os {
         this->registerLogger(this);
     }
 
+
+    static char logBuffers[FW_LOG_TEXT_BUFFER_DEPTH][FW_LOG_TEXT_BUFFER_SIZE];
```
Comment (kevin-f-ortega) [inline fprime-vxworks/VxWorks/Log.cpp:16]: One more comment:

`FW_LOG_TEXT_BUFFER_DEPTH` seems to be undefined. I looked through the fprime repo and could not find it.

Comment (timcanham) [reply fprime-vxworks/VxWorks/Log.cpp:16]: I put it in the `FpConfig.hpp` for Sabertooth:

```
// Define the depth of the Os::Logger circular buffer - may not be used on all platforms
#ifndef FW_LOG_TEXT_BUFFER_DEPTH
#define FW_LOG_TEXT_BUFFER_DEPTH              100   //!< Max size of string for text log message
#endif
```

The idea is that a project could define the depth of the FIFO depending on their memory requirements.

Comment (kevin-f-ortega) [reply fprime-vxworks/VxWorks/Log.cpp:16]: Gotcha. Thank you. LGTM

**Pull request merged**